### PR TITLE
Ruby ART19 Onboarding exercise

### PR DESCRIPTION
**Pull request by mghorb**

### CODE REVIEW

File: app/controllers/api/v1/tags_controller.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
+class Api::V1::TagsController < ApplicationController
+
+  def index
+    @tags = Tag.joins(:articles).where(id: params[:id])
+    render json: @tags, status: 200
```
Comment (demersus) [inline app/controllers/api/v1/tags_controller.rb:5]: This index endpoint shouldn't care about the article id.  Just return all the tags.
- [ ] GET /tags?q=
  - the `q`  param should be used to do a partial (ILIKE) text match on the tag name

File: app/controllers/api/v1/tags_controller.rb [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  def new
```
Comment (demersus) [inline app/controllers/api/v1/tags_controller.rb:8]: This action isn't needed for a json api.  The client application should construct the new tag and save it via `POST /tags`

File: app/controllers/api/v1/tags_controller.rb [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    @tag = Tag.new
+  end
+
+  def show
+    @tags = Tag.all
```
Comment (demersus) [inline app/controllers/api/v1/tags_controller.rb:13]: Show actions should return only a single resource.

File: app/controllers/api/v1/tags_controller.rb [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+    end
+  end
+
+  def create
+    @tag = Tag.new(tag_params)
+
+    if @tag.save
+      render json: @tag, status: 200
+    else
+      render json: {error: "Tag not created."}
```
Comment (demersus) [inline app/controllers/api/v1/tags_controller.rb:27]: In this case we'll return the validation errors in the resopnse. As well as a 422 status code.

File: app/controllers/api/v1/tags_controller.rb [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+    end
+  end
+
+  def update
+    @tag = Tag.find(params[:id])
+
+    if @tag.update(tag_params)
+      render json: @tag, status: 200
+    else
+      render json: {error: "Tag cannot be updated."}
```
Comment (demersus) [inline app/controllers/api/v1/tags_controller.rb:37]: Same here.  For a proper json api you'll need to return a 422 status code and validation error messages.

File: app/controllers/api/v1/tags_controller.rb [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+    end
+  end
+
+  def destroy
+    @tag = Tag.find(params[:id])
+    @tag.destroy
+    render json: "Tag deleted."
```
Comment (demersus) [inline app/controllers/api/v1/tags_controller.rb:44]: Our typical pattern here would be to return an empty body with a 204 status code.

File: app/controllers/comments_controller.rb [hunk 7]
```diff
@@ -0,0 +1,39 @@
+class CommentsController < ApplicationController
+
+  def create
+    @article = Article.find(params[:article_id])
+    @comment = @article.comments.create(comment_params)
+
+    VerifyComment.perform_later(@comment)
```
Comment (demersus) [inline app/controllers/comments_controller.rb:7]: Please trigger the verification worker from an active record callback in the model.  If a comment would be created through another entry-point, we'd want the same verification process to happen.

File: app/controllers/comments_controller.rb [hunk 8]
```diff
@@ -0,0 +1,39 @@
...
+    @article = Article.find(params[:article_id])
+    @comment = @article.comments.find(params[:id])
+  end
+
+  def update
+    @article = Article.find(params[:article_id])
+    @comment = @article.comments.find(params[:id])
+
+    if @comment.update(comment_params)
+      VerifyComment.perform_later(@comment)
```
Comment (demersus) [inline app/controllers/comments_controller.rb:21]: Same here with an ActiveRecord callback. 
Also, the comment needs to be changed to the 'submitted' state on every update.  So you could trigger the VerifyComment worker after save, only when in the 'submitted' state.

File: app/controllers/verify_comment.rb [hunk 9]
```diff
@@ -0,0 +1,13 @@
+class VerifyComment < ApplicationJob
+  queue_as :default
+
+  TRIGGER_WORDS = ['hate','bad','boo', ':(']
```
Comment (demersus) [inline app/controllers/verify_comment.rb:4]: We'll freeze constants.
```suggestion
  TRIGGER_WORDS = ['hate','bad','boo', ':('].freeze
```

Another thing you'll notice is that we will put the [magic comment](https://www.mikeperham.com/2018/02/28/ruby-optimization-with-one-magic-comment/) at the top of every ruby file to freeze string literals:
```
# frozen_string_literal: true

class ...
```

File: app/models/concerns/approval.rb [hunk 10]
```diff
@@ -0,0 +1,14 @@
+module Approval
+  extend ActiveSupport::Concern
+
+  APPROVAL_STATUSES = ['submitted', 'approved', 'flagged']
```
Comment (demersus) [inline app/models/concerns/approval.rb:4]:
```suggestion
  APPROVAL_STATUSES = ['submitted', 'approved', 'flagged'].freeze
```

File: db/schema.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    t.integer "tag_id", null: false
+  end
+
+  create_table "comments", force: :cascade do |t|
+    t.string "commenter"
+    t.text "body"
+    t.integer "article_id", null: false
+    t.datetime "created_at", precision: 6, null: false
+    t.datetime "updated_at", precision: 6, null: false
+    t.string "status"
```
Comment (demersus) [inline db/schema.rb:34]: default status should be 'submitted'

File: spec/controllers/articles_controller_spec.rb [hunk 12]
```diff
@@ -0,0 +1,45 @@
...
+
+    after(:all) do
+      @article.destroy
+    end
+
+    it 'is valid with valid attributes' do
+      expect(@article).to be_valid
+    end
+
+    # Was having trouble with this one - I tried (with 'shoulda-matchers' lib)
```
Comment (mghorb) [inline spec/controllers/articles_controller_spec.rb:18]: describe '#verify_comment' do

Comment (mghorb) [inline spec/controllers/articles_controller_spec.rb:18]: @demersus Tried to explain what I was having trouble here with, feel free to PM me

File: spec/controllers/articles_controller_spec.rb [hunk 13]
```diff
@@ -0,0 +1,45 @@
...
+    # it 'checks uniqueness of title' do
+    #   expect(@article).to validate_uniqueness_of(:title)
+    # end
+    #
+    # which gave me:
+    # NoMethodError: undefined method `validate_uniqueness_of'
+    #
+    #
+    # And I tried:
+    # it 'checks uniqueness of title' do
```
Comment (demersus) [inline spec/controllers/articles_controller_spec.rb:30]: You can use the block form with `except`:
```ruby
let(:article) { FactoryBot.create(:article) }

it 'validates the uniqueness of title' do
  expect { FactoryBot.create(:article, title: article.title) }.to raise_error 'Validation failed: Title has already been taken'
end
```

This allows RSpec to defer execution until you chain the assertion from it.

File: spec/controllers/comments_controller_spec.rb [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    end
+
+    it 'is valid with valid attributes' do
+      @comment = @article.comments.create(commenter: 'Joe', body: 'blahblahblah', status: 'submitted')
+      expect(@comment).to be_valid
+    end
+
+    it 'flags comment when trigger word is found' do
+      @comment = @article.comments.create(commenter: 'Jill', body: 'bad comment', status: 'submitted')
+      VerifyComment.new.perform(@comment)
```
Comment (demersus) [inline spec/controllers/comments_controller_spec.rb:21]: This assertion should be in the spec for VerifyComment. 

Here, I would use rspec mocks to check if the article is queued for verification:

```ruby
before { allow(VerifyComment).to receive(:perform_async) }

it 'will queue a VerifyComment worker when created' do
  article = FactoryBot.create(:article)
  expect(VerifyComment).to have_received(:perform_async).with(article.id)
end
```

File: spec/factories.rb [hunk 15]
```diff
@@ -0,0 +1,13 @@
+FactoryBot.define do
+
+  factory :article do
+    title { 'Article x' }
```
Comment (demersus) [inline spec/factories.rb:4]: In your block here, you can get an argument for a counter.  Then you can add that to your title to ensure multiple created have unique titles.

`title { |i| "Aricle #{i}" }`

File: spec/controllers/comments_controller_spec.rb [hunk 16]
```diff
@@ -0,0 +1,26 @@
+require 'rails_helper'
```
Comment (demersus) [inline spec/controllers/comments_controller_spec.rb:1]: These appear to be model specs.  I'd put this in spec/models/comment_spec.rb 
Similarly with the articles_controller_spec.  You are testing the model.  Those assertions belong in a model spec.

You'll need to add request specs to cover the request over http through the response.   Controller specs are then only used if there's unit-level complexity to test.   Otherwise, request specs are generally enough for standard API controllers.

File: db/migrate/20211207202435_create_join_table_articles_tags.rb [hunk 17]
```diff
@@ -0,0 +1,8 @@
+class CreateJoinTableArticlesTags < ActiveRecord::Migration[6.1]
+  def change
+    create_join_table :articles, :tags do |t|
```
Comment (demersus) [inline db/migrate/20211207202435_create_join_table_articles_tags.rb:3]: Let me give you a hint here.  If you turn this into an actual model, you can get a counter cache macro to cover the tag counter acceptance criteria.

https://guides.rubyonrails.org/association_basics.html#options-for-belongs-to-counter-cache

**Review by demersus — CHANGES_REQUESTED** [review]
Looks like you're making good progress here.

I'd like to see you continue to improve this with request specs, and specs around the comment model's statuses and changing body.  All resources will need to respond to JSON format also.

File: app/workers/verify_comment.rb [hunk 18]
```diff
@@ -0,0 +1,15 @@
+# frozen_string_literal: true
+
+class VerifyComment < ApplicationJob
+  queue_as :default
+
+  TRIGGER_WORDS = ['hate','bad','boo', ':(']
+
+  def perform(comment)
+    if TRIGGER_WORDS.any? { |s| comment.body.include? s}
+      comment.update_column(:status, 'flagged')
```
[the remark this answers is not in the document]
Comment (mghorb) [reply app/workers/verify_comment.rb:10]: Oh yeah so I had some trouble with this which is why I used `update_column`, basically like a stack overflow. `after_save` will trigger after update and will trigger this function, which will trigger the `comment.update`  which will trigger `after_save`..... 

If you have another solution to avoid this infinite loop please lmk, I spent a while trying to do this so I found `update_column` what worked for me

File: spec/controllers/articles_controller_spec.rb [hunk 19]
```diff
@@ -0,0 +1,45 @@
+require 'rails_helper'
+
+RSpec.describe ArticlesController, type: :controller do
+
+  describe '#verify_article' do
+    before(:all) do
+      @article = FactoryBot.create(:article)
```
[the remark this answers is not in the document]
Comment (mghorb) [reply spec/controllers/articles_controller_spec.rb:7]: Sorry just need a little more context on where specifically this would go? because otherwise wouldnt @article in the instances be null?

File: spec/controllers/comments_controller_spec.rb (hunk 14 above)
Comment (mghorb) [reply spec/controllers/comments_controller_spec.rb:21]: Ok so i've tried `.to receive(:perform_async)`, which gives me: 
```
 Failure/Error: before { allow(VerifyComment).to receive(:perform_async) }
       VerifyComment does not implement: perform_async
     # ./spec/workers/verify_comment_spec.rb:10:in `block (3 levels) in <main>'
```

and also tried `.to receive(:perform_async)`:
```
(VerifyComment (class)).perform_later(50)
           expected: 1 time with arguments: (50)
           received: 0 times
```
any thoughts?

Comment (demersus) [reply spec/controllers/comments_controller_spec.rb:21]: I see now that you are inheriting from ActiveJob instead of Sidekiq::Worker.  In this case, the method would be: #perform_later.  Instead of #perform_async

Comment (demersus) [reply spec/controllers/comments_controller_spec.rb:21]: We recommend sidekiq directly in our systems:  https://andycroll.com/ruby/use-sidekiq-directly-not-through-active-job/

This is fine though for this application.  The general concepts are the same.

Comment (mghorb) [reply spec/controllers/comments_controller_spec.rb:21]: OK thanks! let me try again

Comment (mghorb) [reply spec/controllers/comments_controller_spec.rb:21]: Oh yeah I did try :perform_later, gave me the this exception: 
```
(VerifyComment (class)).perform_later(50)
           expected: 1 time with arguments: (50)
           received: 0 times
```

Let me try to investigate why ..
